Use lower shadow normal bias for distant directional shadow splits (reverted)

Calinou opened the pull request:
Follow-up to https://github.com/godotengine/godot/pull/48776.

This reduces the visual discrepancy between shadow splits when Normal Bias is greater than 0. Shadow acne at a distance may be slightly more present, but if the bias and normal bias values are tuned to avoid acne up close, acne will usually not be present in distant splits.

Tested on desktop (PCSS + PCF) and mobile (PCF) backends, with and without Blend Splits enabled. Performance is identical compared to before.

**Testing project:** [test_shadow_splits.zip](https://github.com/user-attachments/files/18131989/test_shadow_splits.zip)

## Preview

### Stills (from the MRP)

#### Blend Splits disabled

Before | After
-|-
![Screenshot_20241213_222624](https://github.com/user-attachments/assets/e6612f2d-b566-4192-b4fd-b22d9979d6a1) | ![Screenshot_20241213_222851](https://github.com/user-attachments/assets/953f8e3d-30a5-45de-be87-648b084562e0)

#### Blend Splits enabled

Before | After
-|-
![Screenshot_20241213_222631](https://github.com/user-attachments/assets/9aff1622-604f-4d59-b09f-36630b19d76c) | ![Screenshot_20241213_222856](https://github.com/user-attachments/assets/ec24adaf-c73d-4fab-99fd-03e6d22ca31f)

### In motion

*Normal Bias was increased to 10 to exagerate the difference. Usually, the difference isn't as pronounced, but it can still be significant when using Normal Bias set to 2 as done in https://github.com/godotengine/godot/pull/55757.*

### Before

https://user-images.githubusercontent.com/180032/162993467-b17af410-587a-4f1d-b954-d213986f61fe.mp4

### After

https://user-images.githubusercontent.com/180032/162993523-77409f26-1b6e-433a-b516-99f0d7e70b3a.mp4

mrjustaguy commented:
This causes Shadow Acne, especially the lower the value of Split 1 is. This also hinders current mitigation of issue that prompted https://github.com/godotengine/godot-proposals/issues/4387 making it worse

Here's a Test Scene, open in current master (with the blur PR merged) and this PR
[Test.zip](https://github.com/godotengine/godot/files/8474434/Test.zip)

mrjustaguy commented:
Mitigations for the test example were found, shrinking objects slightly or using https://github.com/godotengine/godot-proposals/issues/4517 techniques (reversing face culling and shrinking slightly) helps, and generally as the Acne becomes visible only on faces viewed at grazing angles (by the light) it generally isn't actually that much of a problem.

Still The new behavior should be documented with mitigations, for those that do run into these edge cases if this is merged, until the issue gets resolved, if it ever does.

cosmicchipsocket commented:
> This causes Shadow Acne, especially the lower the value of Split 1 is (changing other splits also changes things, but not as much as Split 1 seems to). This also hinders current mitigation of issue that prompted [godotengine/godot-proposals#4387](https://github.com/godotengine/godot-proposals/issues/4387) making it worse

Perhaps this should be exposed as a tweakable then? I've found that setting high normal bias values magically erases peter panning while allowing bias to be set higher, meaning that shadow acne is also defeated. The shadow improvements are specifically why I'd like to migrate to Godot 4, as reverse face culling introduces multiple issues that can't all be resolved by just changing the geometry. And for my application, I can get away with lower bias as I will be opting for little to no shadow filtering, meaning less shadow acne.

mrjustaguy commented:
Is this PR going to get looked at again anytime soon?

Calinou committed:
Use lower shadow normal bias for distant directional shadow splits

This reduces the visual discrepancy between shadow splits when
Normal Bias is greater than 0. Shadow acne at a distance may
be slightly more present, but if the bias and normal bias values are
tuned to avoid acne up close, acne will usually not be present in
distant splits.

Calinou commented:
Rebased and tested again on Forward+/Mobile with and without Blend Splits (in various split configurations), it works as expected. I've incorporated @clayjohn's suggestion so that no shader changes are required.

clayjohn commented on servers/rendering/renderer_rd/storage_rd/light_storage.cpp:
```diff
@@ -714,7 +714,13 @@ void LightStorage::update_light_buffers(RenderDataRD *p_render_data, const Paged
 						float bias_scale = light_instance->shadow_transform[j].bias_scale * light_data.soft_shadow_scale;
 						light_data.shadow_bias[j] = light->param[RS::LIGHT_PARAM_SHADOW_BIAS] / 100.0 * bias_scale;
-						light_data.shadow_normal_bias[j] = light->param[RS::LIGHT_PARAM_SHADOW_NORMAL_BIAS] * light_instance->shadow_transform[j].shadow_texel_size;
+						if (j >= 1) {
```
There is no need for the branch here since `light_data.shadow_split_offsets[0] / split` is `1.0` when `j` is `0`

clayjohn reviewed (commented):
Discussed in the rendering meeting. 

@badsectoracula Will test this on another scene and see how it looks. But ultimately we are happy to merge this for 4.5 and see if anyone complains

clayjohn commented:
@badsectoracula Did you ever get a chance to test this?

badsectoracula commented:
In pretty much all the tests i could think of this PR was fine and didn't really show much of a visual difference. The biggest difference was when i tried with a large area, expanded distance and a directional light that casts long shadows in which case this PR was actually an improvement. Check the two screenshots below, the first one is the current master and the second is master with this PR applies. The shadow shapes look more "correct" with the PR to me and in one case (red arrow) the current behavior in master "breaks" the shape whereas with the PR it looks correct:

Current master:
![image](https://github.com/user-attachments/assets/f14a0ee0-d69b-411f-90a1-f325a2e672b7)

With the PR applied:
![image](https://github.com/user-attachments/assets/8ed7c995-f4f3-4b58-900a-9108a08a25ff)

Repiteo commented:
Thanks!
